src/coin_test/data/dataset_generator.py

```diff
@@ -210,3 +214,396 @@ class ReturnsDatasetGenerator(StitchedChunkDatasetGenerator):
+            extreme_series = close_series.combine(open_series, max, fill_value=0)
+        else:
+            extreme_series = close_series.combine(open_series, min, fill_value=0)
+
+        # Remove all series index before calculation.
+        series_to_pct_change = series_to_pct_change.reset_index(drop=True)
+        extreme_series = extreme_series.reset_index(drop=True)
+
+        pct_change_series = (series_to_pct_change - extreme_series) / extreme_series
+
+        return pct_change_series
+
+    @staticmethod
+    def from_pct_change_based_on_extreme_bar_open_and_close(
+        pct_change_series: pd.Series,
+        extreme: Literal["max", "min"],
+        open_series: pd.Series,
+        close_series: pd.Series,
+    ) -> pd.Series:
+        """Calculates values from percent change and min/max of Open and Close.
+
+        Args:
+            pct_change_series: a series of percent change to turn into values
+            extreme: selects either min or max when calculating extreme series
+            open_series: series of open price values
+            close_series: series of close price values
+
+        Returns:
+            series: extreme_series + extreme_series * pct_change_series
+
+        Raises:
+            ValueError: open_series should be same length as close_series.
+            ValueError: pct_change_series should be same length as open_series.
+        """
+        if len(close_series) != len(open_series):
+            raise ValueError("open_series should be same length as close_series.")
+
+        if len(pct_change_series) != len(open_series):
+            raise ValueError("pct_change_series should be same length as open_series.")
+
+        if extreme == "max":
+            extreme_series = close_series.combine(open_series, max, fill_value=0)
+        else:
+            extreme_series = close_series.combine(open_series, min, fill_value=0)
+
+        # Remove all series index before calculation.
+        pct_change_series = pct_change_series.reset_index(drop=True)
+        extreme_series = extreme_series.reset_index(drop=True)
+
+        return extreme_series + extreme_series * pct_change_series
+
+    @staticmethod
+    def from_pct_change_based_on_starting_value(
+        pct_change_series: pd.Series,
+        starting_price: float,
+    ) -> pd.Series:
+        """Generate synthetic series from GARCH model fit to univariate data."""
+        # Values are one row longer than by-item percent changes.
+        num_rows = len(pct_change_series) + 1
+        # Remove all series index before calculation.
+        pct_change_series = pct_change_series.reset_index(drop=True)
+
+        # Transform generated percent change values to price values.
+        synthetic_series = pd.Series(0, index=list(range(num_rows)))
+        # Intialize series_0 from starting value.
+        synthetic_series[0] = starting_price
+        # set series_j = series_j-1 + series_j-1 * pct_change_j-1 / 100.
+        for j in range(1, num_rows):
```

Comment:
Would it be possible to do this with pandas operations? Maybe an apply over a rolling window or something? Not strictly necessary though.

Reply:
I found this [stackoverflow post](https://stackoverflow.com/questions/34855859/is-there-a-way-in-pandas-to-use-previous-row-value-in-dataframe-apply-when-previ) for handling setting the value when the inputs to the calculation are not known in the beginning. I could use the apply function? Will get back to you on this.

Reply:
Hmm, it does seem hard. Not a priority at all - I think its fine to leave as is.

Reply:
From the stackoverflow post, the solution using apply also uses a global variable. Not standard but faster on average?
```python
import pandas as pd
import numpy as np

data = np.array([[10, 2, 10, 10],
                 [10, 3, 60, 100],
                 [np.nan] * 4,
                 [10, 22, 280, 250]]).T
idx = pd.date_range('20150131', end='20150203')
df = pd.DataFrame(data=data, columns=list('ABCD'), index=idx)
df
               A    B     C    D
 =================================
 2015-01-31    10   10    NaN  10
 2015-02-01    2    3     NaN  22 
 2015-02-02    10   60    NaN  280
 2015-02-03    10   100   NaN  250

def calculate(mul, add):
    global value
    value = value * mul + add
    return value

value = df.loc['2015-01-31', 'D']
df.loc['2015-01-31', 'C'] = value
df.loc['2015-02-01':, 'C'] = df.loc['2015-02-01':].apply(lambda row: calculate(*row[['A', 'B']]), axis=1)
df
               A    B     C     D
 =================================
 2015-01-31    10   10    10    10
 2015-02-01    2    3     23    22 
 2015-02-02    10   60    290   280
 2015-02-03    10   100   3000  250
```